specification/hugr.md

```diff
@@ -665,68 +672,103 @@ See [Type System](#type-system) for more on Resources.
+      outputs: [[null, F64, 1]]
+  - name: ArrayConcat
+    description: "Concatenate two arrays. Resource provides a compute_signature implementation."
     args:
-      - matrix: List(List(List(F64))))
+      - t: Type  # Classic or Quantum
+      - i: U64
+      - j: U64
+    # inputs could be: Array<i>(t), Array<j>(t)
+    # outputs would be, in principle: Array<i+j>(t)
+    # - but default type scheme interpreter does not support
+    # No signature block: will look up a compute_signature based on resource name
+```
 
-- name: MyResource
-  operations:
-  - name: MyCustom
-    description: "Custom op defined by a program"
-    resource_reqs: [MyGates] # Depend on operations defined in the other module
-    inputs: [[null, Q], [null, Q], [param, F64]]
-    outputs: [[null, Q], [null, Q]]
+The declaration of the `args` uses a language that is a distinct, simplified
+form of the [Type System](#type-system) - writing terminals that appear in the YAML in quotes,
+the value of each member of `args` is given by the following production:
 ```
+TypeParam ::= "Type" | "ClassicType" | "F64" | "U64" | "I64" | "Opaque"(name, ...) | "List"(TypeParam)
+```
+
+The structure of the `signature` block is as follows.
+* It may contain a `function` of string type. This is used to look up a `compute_signature`
+  among implementations that have self-registered (e.g. impl's of a Rust trait).
+* If the `signature` block is present but no `function` is provided, the default
+  type-scheme-interpreter is used
+* Any *other* members of the `signature` block (i.e. properties of the OpDef, the same for
+  all operation nodes) are passed to `compute_signature` too, along with the OpDef name, the
+  `misc` data (also passed to other compiler passes) and the node's values for the `args`
+   * For the default type scheme interpreter, the `signature` block will have to include
+     both `inputs` and `outputs`
+* If the signature block is absent, this is equivalent to a signature block containing only
+  `function: <name-of-resource>`.
 
 Reading this format into Rust is made easy by `serde` and
 [serde\_yaml](https://github.com/dtolnay/serde-yaml) (see the
 Serialization section). It is also trivial to serialize these
 definitions in to the overall HUGR serialization format.
 
-Note the required `name`, `description`. `inputs` and `outputs` fields,
-the last two defining the signature of the operation, and optional
-parameter names as metadata. The optional `misc` field is used for
-arbitrary YAML, which is read in as-is (into the `serde_yaml Value`
-struct). The data held here can be used by compiler passes which expect
-to deal with this operation (e.g. a pass can use the `basis` information
-to perform commutation). The optional `args` field can be used to
-specify the types of parameters to the operation - for example the
-matrix needed to define an SU2 operation.
+Note the only required fields are `name` and `description`. `signature` is optional, as
```

Comment:
It does feel here that `misc` and `signature` are asking to be combined; they are both almost-arbitrary data constant across nodes using the definition.